h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -17,14 +17,81 @@
    *
    * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained. Only Frame and Model keys are accepted.
    */
-  public static void retain(final Key[] retainedKeys){
-    final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
-    retainedSet.addAll(Arrays.asList(retainedKeys));
+  public static void retain(final Key[] retainedKeys) {
+    final Set<Key> retainedKeysSet = new HashSet<>(retainedKeys.length);
+    retainedKeysSet.addAll(Arrays.asList(retainedKeys));
     // Frames and models have multiple nested keys. Those must be extracted and kept from deletion as well.
-    extractNestedKeys(retainedSet);
+    extractNestedKeys(retainedKeysSet);
+    Set<Value> removedKeys = collectUniqueKeys();
+    removeValues(removedKeys, retainedKeysSet);
+  }
+
+  /**
+   * Collects a {@link Set} of keys available cluster-wide
+   *
+   * @return
+   */
+  private static final Set<Value> collectUniqueKeys() {
+    final Value[] collectedValues = new CollectValuesTask()
+            .doAllNodes()
+            ._collectedValues;
+    final Set<Value> clusterValues = new HashSet<>(collectedValues.length);
```

Comment:
Why is this necessary?

Reply:
It's not, you're right. Removed. Now there is an array.